src/worker.cc

```diff
@@ -1011,31 +1122,75 @@ std::string jsonify_error(const valhalla_exception_t& exception, Api& request) {
   err_stat->set_value(1);
   err_stat->set_type(count);
 
+  // pbf format output
+  if (request.options().format() == Options::pbf) {
+    return request.SerializeAsString();
+  }
+
+  // json
   return body.str();
 }
 
+void ParseApi(const std::string& request, Options::Action action, valhalla::Api& api) {
+  // if its a pbf request we only want to clear the stuff we are going to eventually set
+  if (api.has_options() && request.empty()) {
+    api.clear_trip();
+    api.clear_directions();
+    api.clear_status();
+    api.clear_info();
+  } // otherwise blank it all just in case
+  else {
+    api.Clear();
+  }
+
+  // maybe parse some json
+  api.mutable_options()->set_action(action);
+  auto document = from_string(request, valhalla_exception_t{100});
+  from_json(document, *api.mutable_options());
+}
+
 #ifdef HAVE_HTTP
 void ParseApi(const http_request_t& request, valhalla::Api& api) {
-  api.Clear();
-
   // block all but get and post
   if (request.method != method_t::POST && request.method != method_t::GET) {
     throw valhalla_exception_t{101};
   };
 
+  // set the action
+  api.Clear();
+  auto& options = *api.mutable_options();
+  Options::Action action;
+  if (!request.path.empty() && Options_Action_Enum_Parse(request.path.substr(1), &action)) {
+    // if this is a pbf request, its action is overridden by the request path
+    options.set_action(action);
+  }
+
+  // if its a protobuf mime go with that
```

Comment:
if we see you are telling us its protobuf then we abort looking at query params or trying to parse json at all and just assume it has a request body and that that body is protobuf bytes